[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98105

Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

swolchok committed:
[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)

Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok commented:
previously: #83707

swolchok commented on c10/util/typeid.h:
```diff
@@ -520,6 +520,7 @@ class C10_API TypeMeta final {
 
   static uint16_t existingMetaDataIndexForType(TypeIdentifier identifier);
 
+ public:
```
I debated adding _DO_NOT_CALL_DIRECTLY to the name of addTypeMetaData, but that didn't seem very PyTorchy.

ezyang replied:
meh, I don't mind haha

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok commented:
legit windows failures that I missed -- `error LNK2019: unresolved external symbol "unsigned short const caffe2::detail::std_string_metadata_index"` and similar

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok commented:
@pytorchbot cancel

weiwangmeta commented:
[I manually cancelled via this job: ](https://github.com/pytorch/pytorch/actions/runs/4641203578)

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok commented:
macos-12-py3-arm64 failure looks unrelated. will try having it rebase before merge

swolchok commented:
@pytorchbot merge --rebase

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)"


Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -f 'already had green signal; simply re-exported to get rid of meta internal-only changes notice'
